chalk-integration/src/lowering.rs

```diff
@@ -915,63 +954,67 @@ trait LowerTy {
                     if k.binders.len() > 0 {
                         Err(RustIrError::IncorrectNumberOfTypeParameters {
                             identifier: name,
                             expected: k.binders.len(),
                             actual: 0,
-                        })?
+                        })
                     } else {
                         Ok(chalk_ir::TyData::Apply(chalk_ir::ApplicationTy {
-                            name: chalk_ir::TypeName::TypeKindId(id.into()),
+                            name: chalk_ir::TypeName::Struct(id),
                             parameters: vec![],
                         })
                         .intern())
                     }
                 }
-                NameLookup::Parameter(d) => Ok(chalk_ir::TyData::BoundVar(d).intern()),
+                TypeLookup::Parameter(d) => Ok(chalk_ir::TyData::BoundVar(d).intern()),
             },
 
-            Ty::Dyn { ref bounds } => Ok(chalk_ir::TyData::Dyn(env.in_binders(
-                // FIXME: Figure out a proper name for this type parameter
-                Some(chalk_ir::ParameterKind::Ty(intern(FIXME_SELF))),
-                |env| {
-                    Ok(bounds
-                        .lower(env)?
-                        .iter()
-                        .flat_map(|qil| {
-                            qil.into_where_clauses(chalk_ir::TyData::BoundVar(0).intern())
-                        })
-                        .collect())
-                },
-            )?)
+            Ty::Dyn { ref bounds } => Ok(chalk_ir::TyData::Dyn(chalk_ir::BoundedTy {
+                bounds: env.in_binders(
+                    // FIXME: Figure out a proper name for this type parameter
+                    Some(chalk_ir::ParameterKind::Ty(intern(FIXME_SELF))),
+                    |env| {
+                        Ok(bounds
+                            .lower(env)?
+                            .iter()
+                            .flat_map(|qil| {
+                                qil.into_where_clauses(chalk_ir::TyData::BoundVar(0).intern())
+                            })
+                            .collect())
+                    },
+                )?,
+            })
             .intern()),
 
-            Ty::Opaque { ref bounds } => Ok(chalk_ir::TyData::Opaque(env.in_binders(
-                // FIXME: Figure out a proper name for this type parameter
-                Some(chalk_ir::ParameterKind::Ty(intern(FIXME_SELF))),
-                |env| {
-                    Ok(bounds
-                        .lower(env)?
-                        .iter()
-                        .flat_map(|qil| {
-                            qil.into_where_clauses(chalk_ir::TyData::BoundVar(0).intern())
-                        })
-                        .collect())
-                },
-            )?)
+            Ty::Opaque { ref bounds } => Ok(chalk_ir::TyData::Opaque(chalk_ir::BoundedTy {
+                bounds: env.in_binders(
```

Comment:
This code is straight duplicated. Maybe this is worth a helper function?

Reply:
Actually the `Opaque` code should be removed, I recently realized this is the wrong way to model it (that gist I've been preparing talks a bit about it). I'll do that now.